include/nccl_ofi_memcheck.h

```diff
@@ -52,7 +52,7 @@ _Static_assert(MEMCHECK_REDZONE_SIZE % MEMCHECK_GRANULARITY == 0,
  * @param size
  *   The size of the memory area.
  */
-static inline void nccl_net_ofi_mem_defined(void *data, size_t size);
+static inline void nccl_net_ofi_mem_defined(uintptr_t data, size_t size);
```

Comment:
The "correct" modern C++ way to do this is described well in this SO answer: https://stackoverflow.com/a/45461976

But until we've actually moved fully to C++, using std::byte* is off the table, so this is a reasonable middleground right now.

Reply:
I should also note, I have loose plans to replace the entirety of the freelist implementation with `std::pmr::memory_resource`/`cuda::mr::resource`/`rmm::mr::pinned_host_memory_resource`, so I don't think this file is going to survive through the end of the year.

Reply:
I'm dubious of changing to a bunch of C++ interfaces, but I still don't understand this change.  There's no reason for this to be a `uintptr_t`.

Reply:
arithmetic on pointer to void is [undefined behavior in both C and C++](https://stackoverflow.com/a/3524270) and the difference is that C will accept it and most implementations do the right thing, whereas it simply won't compile in C++.

"uintptr_t  is an unsigned integer type such that any valid (void *) value can be converted to this type and then converted back." - man 3 uintptr_t, so this is a reasonable replacement. This should not be controversial. I'm not trying to sneak something past you here, I do not care whatsoever if this API is defined over void* or uintptr_t, I'm trying to get the code to build.

Reply:
> I'm dubious of changing to a bunch of C++ interfaces, but I still don't understand this change. There's no reason for this to be a uintptr_t.

The general value proposition for std::pmr::memory_resource and the nvidia projects building atop it is this: if you know at the time of defining or instantiating the struct what its pinning/alignment/backing storage requirements are, you can put `using allocator_type = XYZ;` in the struct, or pass an allocator to the constructor, and you will never have to think about it again. It can be backed by a thread-local arena, or backed by a buffer on the stack, or it can be backed by the default global new/delete allocator, or it can be backed by accelerator memory through cuda APIs, or it can be backed by a custom hugepage allocator, etc. But the point is you define it once and then you never think about it again. You can blissfully write vanilla C++ with these structs similar to if you were using `new` and `delete` or smart pointers, as if they are standard heap-backed objects, with all of these pinning and alignment and storage guarantees abstracted away completely. Bloomberg had implemented this interface internally and pushed to standardize it ~10 years ago for C++17, so you can imagine it's a mature API at this point that's battle tested for low-latency work where you need arenas and guarantees around when allocation is acceptable and when it's not.

It's also polymorphic, which means you can define allocators in terms of a parent allocator. One cool example from RMM is their [logging allocator adapter](https://docs.rapids.ai/api/rmm/nightly/librmm_docs/memory_resources/#_CPPv4I0EN3rmm2mr24logging_resource_adaptorE), which can wrap an allocator and collect stats for every allocation and free on that allocator. Another neat idea is to replace the upstream on an allocator with the null memory resource, which lets you guarantee that after a given point in the code, any attempt to request more memory will abort.

It's likely to be more efficient than what we currently have with our padding and linked lists, and it's also code that we won't have to write; nvidia and the standard library are vouching for these interfaces.

Reply:
I'm not arguing over doing arithmetic operations on a void pointer, but looking at the interface, why do the cast at the top of the stack, instead of having the right interface and a cast at the bottom?  A void * is the type that makes sense for these interfaces.